src/app/content/highlights/components/CardWrapper.tsx

```diff
@@ -66,23 +67,42 @@ const Wrapper = ({highlights, className, container, highlighter}: WrapperProps)
   }, [cardsHeights]);
 
   const getOffsetsForHighlight = React.useCallback((highlight: Highlight) => {
-    if (offsets.has(highlight.id)) {
-      return assertDefined(offsets.get(highlight.id), 'this has to be defined');
-    } else {
-      const newOffsets = assertDefined(
-        getHighlightOffset(container, highlight),
-        `Couldn't get offsets for highlight with an id: ${highlight.id}`
-      );
-      setOffsets((state) => new Map(state).set(highlight.id, newOffsets));
-      return newOffsets;
-    }
-  }, [container, offsets]);
+    const newOffsets = assertDefined(
+      getHighlightOffset(container, highlight),
+      `Couldn't get offsets for highlight with an id: ${highlight.id}`
+    );
+    setOffsets((state) => new Map(state).set(highlight.id, newOffsets));
+    return newOffsets;
+  }, [container]);
+
+  const checkIfHiddenByCollapsedAncestor = (highlight: Highlight) => {
+    const highlightElement = highlight.elements[0] as HTMLElement;
+    const collapsedAncestor = highlightElement
+      ? highlightElement.closest('[data-type="solution"][aria-expanded="false"]')
+      : null;
+    return Boolean(collapsedAncestor);
+  };
+
+  React.useEffect(() => {
+    setNewCardsPositionsRef.current = () => {
+      const positions = updateCardsPositions(focusedHighlight, highlights, cardsHeights, getOffsetsForHighlight);
+      setCardsPositions(positions);
+    };
+    setNewCardsPositionsRef.current();
+  }, [cardsHeights, focusedHighlight, getOffsetsForHighlight, highlights]);
 
   React.useEffect(() => {
-    const positions = updateCardsPositions(focusedHighlight, highlights, cardsHeights, getOffsetsForHighlight);
-    setCardsPositions(positions);
-    // eslint-disable-next-line react-hooks/exhaustive-deps
-  }, [highlights, cardsHeights, focusedHighlight]);
+    const resizeObserver = new ResizeObserver(() => {
+      assertDefined(
+        setNewCardsPositionsRef.current,
```

Comment:
could the function itself be assigned to the ref outside of the effect? so that its guaranteed to be defined? i don't see why it needs to be defined in the effect